File: app/src/main/java/com/android/malsoftware/techadvisor/RecycleViewSelect/AdapterSelector.java
package com.android.malsoftware.techadvisor.RecycleViewSelect;

import android.annotation.SuppressLint;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import androidx.annotation.NonNull;
import androidx.databinding.DataBindingUtil;
import androidx.recyclerview.selection.*;
import androidx.recyclerview.widget.RecyclerView;

import com.android.malsoftware.techadvisor.BaseItemViewModel;
import com.android.malsoftware.techadvisor.DescriptionsPresets;
import com.android.malsoftware.techadvisor.MillDetailValues;
import com.android.malsoftware.techadvisor.R;
import com.android.malsoftware.techadvisor.databinding.MillItemDefaultBinding;

import java.util.List;

public class AdapterSelector extends RecycleViewScroll.Adapter<AdapterSelector.SelectorHolder> {

    private final String TAG = "AdapterSelector";
    private SelectionTracker<String> mTracker;
    private DescriptionsPresets mDescriptionsPresets;
    private MillDetailValues Array;
    private List<String> mStrings;

    public AdapterSelector(DescriptionsPresets descriptionsPresets, MillDetailValues array){
        mDescriptionsPresets = descriptionsPresets;
        this.Array = array;
    }

    public void setTracker(SelectionTracker<String> tracker) {
        mTracker = tracker;
    }

    public void setStringKeys(List<String> Keys) {
        mStrings = Keys;
    }

    @NonNull
    @Override
    public SelectorHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        MillItemDefaultBinding millItemDefaultBinding = DataBindingUtil.
                        inflate(inflater, R.layout.mill_item_default, parent, false);
        return new SelectorHolder(millItemDefaultBinding);
    }

    @Override
    public void onBindViewHolder(@NonNull SelectorHolder holder, int position) {
        String item = mStrings.get(position);
        if (mTracker != null) holder.bind(position, Array, mTracker.isSelected(item));
    }

    @Override
    public int getItemCount() {
        return mStrings == null ? 0 : mStrings.size();
    }

    @Override
    public long getItemId(int position) {
        return mStrings.get(position).hashCode();
    }

    class SelectorHolder extends RecyclerView.ViewHolder{

        private MillItemDefaultBinding mMillItemDefaultBinding;
        private BaseItemViewModel mBaseItemViewModel = new BaseItemViewModel();

        SelectorHolder(MillItemDefaultBinding millItemDefaultBinding) {
            super(millItemDefaultBinding.getRoot());
            mMillItemDefaultBinding = millItemDefaultBinding;
            mMillItemDefaultBinding.setBaseModel(mBaseItemViewModel);
        }

        public SelectorHolder(@NonNull View itemView) {
            super(itemView);
        }

        void bind(int position, MillDetailValues val, boolean activated) {
            if (mMillItemDefaultBinding != null) {

                mBaseItemViewModel.setText(val.getStringFieldValue(position));
                itemView.setActivated(activated);
                mBaseItemViewModel.setDescription(mDescriptionsPresets.
                        getStringDescription(val.getFieldType(position)));
            }
        }

        StringItemDetails getItemDetails() {
            return new StringItemDetails(getAdapterPosition(), mStrings.get(getAdapterPosition()));
        }
    }
}

File: app/src/main/java/com/android/malsoftware/techadvisor/RecycleViewSelect/AdapterSelector.kt
package com.android.malsoftware.techadvisor.RecycleViewSelect

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.databinding.ViewDataBinding
import androidx.recyclerview.selection.*
import androidx.recyclerview.widget.RecyclerView
import com.android.malsoftware.techadvisor.BaseItemViewModel
import com.android.malsoftware.techadvisor.DescriptionsPresets
import com.android.malsoftware.techadvisor.MillDetailValues
import com.android.malsoftware.techadvisor.R
import com.android.malsoftware.techadvisor.databinding.MillItemDefaultBinding

class AdapterSelector(private val mDescriptionsPresets: DescriptionsPresets,
                      private val Array: MillDetailValues) :
        RecyclerView.Adapter<AdapterSelector.SelectorHolder>() {

    private val TAG = "AdapterSelector"
    private lateinit var mTracker: SelectionTracker<String>
    private lateinit var mStrings: List<String>

    fun setTracker(tracker: SelectionTracker<String>) {
        mTracker = tracker
    }

    fun setStringKeys(Keys: List<String>) {
        mStrings = Keys
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): SelectorHolder {
        val inflater = LayoutInflater.from(parent.context)
        val millItemDefaultBinding: MillItemDefaultBinding
                = DataBindingUtil.inflate(inflater, R.layout.mill_item_default, parent,
                false)
        return SelectorHolder(millItemDefaultBinding)
    }

    override fun onBindViewHolder(holder: SelectorHolder, position: Int) {
        val item = mStrings[position]
        mTracker.isSelected(item).run { holder.bind(position, Array, this) }
    }

    override fun getItemCount(): Int {
        return mStrings.size
    }

    override fun getItemId(position: Int): Long {
        return mStrings[position].hashCode().toLong()
    }

    open inner class SelectorHolder(millItemDefaultBinding: MillItemDefaultBinding) :
            RecyclerView.ViewHolder((millItemDefaultBinding as ViewDataBinding).root) {

        private var mMillItemDefaultBinding: MillItemDefaultBinding = millItemDefaultBinding
        private val mBaseItemViewModel = BaseItemViewModel()

        val itemDetails: StringItemDetails?
            get() = StringItemDetails(adapterPosition, mStrings[adapterPosition])

        init {
            mMillItemDefaultBinding.baseModel = mBaseItemViewModel
        }

        fun bind(position: Int, millDetailValues: MillDetailValues, activated: Boolean) {
            mBaseItemViewModel.apply {
                setText(millDetailValues.getStringFieldValue(position))
                mDescriptionsPresets.
                        getStringDescription(millDetailValues.
                                getFieldType(position))?.let { setDescription(it)}
            }
            itemView.isActivated = activated
        }
    }
}